File: Day #1585/solution.js
// Day 1585: Sieve of Eratosthenes + incremental infinite prime generator.
// Sieve marks composites up to N; generator yields primes indefinitely via a sieve map.
// Time: O(N log log N) sieve; Space: O(N).
"use strict";

function sieve(n) {
  const comp = new Array(n).fill(false);
  const primes = [];
  for (let i = 2; i < n; i++) {
    if (!comp[i]) {
      primes.push(i);
      for (let j = i * i; j < n; j += i) comp[j] = true;
    }
  }
  return primes;
}

function* primeGenerator() {
  const composites = new Map();
  let n = 2;
  while (true) {
    if (!composites.has(n)) {
      yield n;
      composites.set(n * n, [n]);
    } else {
      for (const p of composites.get(n)) {
        if (!composites.has(n + p)) composites.set(n + p, []);
        composites.get(n + p).push(p);
      }
      composites.delete(n);
    }
    n++;
  }
}

console.log("Primes < 30:", sieve(30).join(" "));
const gen = primeGenerator();
const first10 = [];
for (let i = 0; i < 10; i++) first10.push(gen.next().value);
console.log("First 10 primes:", first10.join(" "));

File: Day #1585/solution.rs
// Day 1585: Sieve of Eratosthenes + incremental infinite prime generator.
// Sieve marks composites up to N; PrimeGen yields primes indefinitely via a sieve map.
// Time: O(N log log N) sieve; Space: O(N).
use std::collections::HashMap;

fn sieve(n: usize) -> Vec<usize> {
    let mut comp = vec![false; n];
    let mut primes = Vec::new();
    for i in 2..n {
        if !comp[i] {
            primes.push(i);
            let mut j = i * i;
            while j < n {
                comp[j] = true;
                j += i;
            }
        }
    }
    primes
}

struct PrimeGen {
    composites: HashMap<u64, Vec<u64>>,
    n: u64,
}

impl Iterator for PrimeGen {
    type Item = u64;
    fn next(&mut self) -> Option<u64> {
        loop {
            let n = self.n;
            self.n += 1;
            match self.composites.remove(&n) {
                None => {
                    self.composites.entry(n * n).or_default().push(n);
                    return Some(n);
                }
                Some(hits) => {
                    for p in hits {
                        self.composites.entry(n + p).or_default().push(p);
                    }
                }
            }
        }
    }
}

fn prime_gen() -> PrimeGen {
    PrimeGen { composites: HashMap::new(), n: 2 }
}

fn main() {
    println!("Primes < 30: {:?}", sieve(30));
    let first10: Vec<u64> = prime_gen().take(10).collect();
    println!("First 10 primes: {:?}", first10);
}
